iib/web/models.py

```diff
@@ -190,6 +209,101 @@ def add_state(self, state, state_reason):
+                self.binary_image_resolved, 'pull_specification', None),
+            'bundles': [bundle.pull_specification for bundle in self.bundles],
+            'from_index': getattr(self.from_index, 'pull_specification', None),
+            'from_index_resolved': getattr(self.from_index_resolved, 'pull_specification', None),
+            'index_image': getattr(self.index_image, 'pull_specification', None),
+            'state': _state_to_json(self.state),
+            'user': getattr(self.user, 'username', None),
+        }
+
+        if verbose:
+            states = [_state_to_json(state) for state in self.states]
+            # Reverse the list since the latest states should be first
+            states = list(reversed(states))
+            rv['state_history'] = states
+
+        return rv
+
+    @classmethod
+    def from_json(cls, kwargs):
+        # Validate all required parameters are present
+        required_params = {'bundles', 'binary_image'}
+        optional_params = {'from_index', 'add_arches', 'cnr_token'}
+        missing_params = required_params - kwargs.keys() - optional_params
+        if missing_params:
+            raise ValidationError(
+                'Missing required parameter(s): {}'.format(', '.join(missing_params))
+            )
+
+        # Don't allow the user to set arbitrary columns or relationships
+        invalid_params = kwargs.keys() - required_params - optional_params
+        if invalid_params:
+            raise ValidationError(
+                'The following parameters are invalid: {}'.format(', '.join(invalid_params))
+            )
+
+        request_kwargs = deepcopy(kwargs)
+
+        # Validate bundles are correctly provided
+        bundles = request_kwargs.pop('bundles')
+        if (
+            not isinstance(bundles, list) or
+            len(bundles) == 0 or
+            not all(isinstance(bundle, str) for bundle in bundles)
+        ):
+            raise ValidationError('"bundles" should be a non-empty array')
+        request_kwargs['bundles'] = [
+            Image.get_or_create(pull_specification=bundle) for bundle in bundles
+        ]
+
+        # Validate binary_image is correctly provided
+        binary_image = request_kwargs.pop('binary_image')
+        if not isinstance(binary_image, str) or not binary_image:
+            raise ValidationError('"binary_image" should be a non-empty string')
+        request_kwargs['binary_image'] = Image.get_or_create(pull_specification=binary_image)
+
+        if not request_kwargs.get('from_index') and not request_kwargs.get('add_arches'):
+            raise ValidationError('One of "from_index" or "add_arches" must be specified')
+
+        # Validate from_index and cnr_token if specified
+        for param in ('from_index', 'cnr_token'):
+            param_value = request_kwargs.pop(param, None)
+            if param_value and not isinstance(param_value, str):
+                raise ValidationError(f'"{param}" must be a string')
```

Comment:
```suggestion
                raise ValidationError(f'"{param}" must be a non-empty string')
```

Reply:
I don't think we should add `non-empty` here because that would imply that it's mandator to specify the param. thoughts?